geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/SortingSimpleFeatureIterator.scala

```diff
@@ -0,0 +1,289 @@
+/***********************************************************************
+ * Copyright (c) 2013-2020 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index.utils
+
+import java.io.{File, FileInputStream, FileOutputStream}
+import java.nio.file.Files
+import java.util.concurrent.atomic.AtomicBoolean
+
+import com.typesafe.scalalogging.LazyLogging
+import org.locationtech.geomesa.features.SerializationOption.SerializationOptions
+import org.locationtech.geomesa.features.SimpleFeatureSerializer
+import org.locationtech.geomesa.features.kryo.KryoFeatureSerializer
+import org.locationtech.geomesa.index.conf.QueryProperties
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureOrdering
+import org.locationtech.geomesa.utils.index.ByteArrays
+import org.locationtech.geomesa.utils.io.{CloseQuietly, Sizable, WithClose}
+import org.opengis.feature.simple.SimpleFeature
+
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Random
+
+/**
+  * In memory sorting of simple features, with optional spill to disk
+  *
+  * @param features unsorted feature iterator
+  * @param sortBy attributes to sort by, in the form: (name, reverse).
+  *               for sort by feature id (e.g. natural sort), use an empty string for name
+  */
+class SortingSimpleFeatureIterator(features: CloseableIterator[SimpleFeature], sortBy: Seq[(String, Boolean)])
+    extends CloseableIterator[SimpleFeature] {
+
+  import SortingSimpleFeatureIterator.{sortInMemory, sortWithSpillover}
+
+  private val closed = new AtomicBoolean(false)
+
+  private lazy val sorted: CloseableIterator[SimpleFeature] = {
+    if (closed.get || !features.hasNext) { features } else {
+      val head = features.next()
+      if (closed.get || !features.hasNext) { CloseableIterator.single(head, features.close()) } else {
```

Comment:
Ok, I'll bite.  Why does the single element iterator get handled separately?

Reply:
it's already sorted ;)

Reply:
I've been in foot races by myself and still come in second...  But yeah, here, that makes sense.